implementations/rust/ockam/ockam/src/system/builder.rs

```diff
@@ -0,0 +1,296 @@
+    C: Send + 'static,
+    M: Message,
+{
+    /// Create an empty SystemBuilder
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Add a new handler to this SystemBuilder
+    ///
+    /// You must at least provide ONE route path (via `.default(addr)`
+    /// or `.condition(cond, addr)`), or else initialisation will
+    /// fail.
+    pub fn add<A, S, H>(&mut self, addr: A, id: S, handler: H) -> HandlerBuilder<'_, C, M>
+    where
+        A: Into<Address>,
+        S: Into<String>,
+        H: SystemHandler<C, M> + Send + 'static,
+    {
+        HandlerBuilder::new(self, addr.into(), id.into(), Box::new(handler))
+    }
+
+    /// Get the address of a previously added SystemHandler
+    pub fn get_addr(&self, id: impl Into<String>) -> Option<Address> {
+        self.inner.get(&id.into()).map(|data| data.addr.clone())
+    }
+
+    /// Specify an entry-point for the soon-to-be-built system
+    pub fn set_entry<A: Into<Address>>(&mut self, addr: A) {
+        self.entry = Some(addr.into());
+    }
+
+    /// Add a new handler to the builder, re-addressing the previous
+    /// handler's "default" route to this one
+    pub fn chain_default<A, H>(
+        &mut self,
+        addr: A,
+        prev_id: impl Into<String>,
+        id: impl Into<String>,
+        handler: H,
+    ) -> HandlerBuilder<'_, C, M>
+    where
+        A: Into<Address>,
+        H: SystemHandler<C, M> + Send + 'static,
+    {
+        self.chain_for("default", addr, prev_id, id, handler)
+    }
+
+    /// Add a new handler to the builder, re-addressing the previous
+    /// handler's named route to this one
+    pub fn chain_for<A, H>(
+        &mut self,
+        rule: impl Into<String>,
+        addr: A,
+        prev_id: impl Into<String>,
+        id: impl Into<String>,
+        handler: H,
+    ) -> HandlerBuilder<'_, C, M>
+    where
+        A: Into<Address>,
+        H: SystemHandler<C, M> + Send + 'static,
+    {
+        let addr = addr.into();
+        let prev_id = prev_id.into();
+        if let Some(ref mut prev) = self.inner.get_mut(&prev_id) {
+            prev.routes.insert(rule.into(), addr.clone());
+        }
+        self.add(addr, id, handler)
+    }
+
+    /// Re-address every system handler in order to clone it
+    ///
+    /// When initialising a worker system builder you setup handlers
+    /// and routes between them.  However when cloning the worker
+    /// system all addresses need to be re-addressed in order to allow
+    /// a second instance of every address to exist on the same node.
+    ///
+    /// This problem may be solved by having scoped addresses in the
+    /// future.
+    ///
+    /// This function assigns a new address to every handler and then
+    /// changes routes that point to the old addresses.  Any route
+    /// that is not in the internal set (i.e. the fin address routes)
+    /// are replaced with the externally provided fin-address.
+    pub fn readdress(&mut self, fin_addr: &Address) {
```

Comment:
Hmm, this functions name reads kind of odd to me readdress -> read-dress or re-address (obviously the latter in this case).

I wonder if we can come up with a better name for it... For now the documentation should be sufficient.